packages/notebook/src/history.ts

```diff
@@ -0,0 +1,365 @@
+    const content = this._filtered[this._cursor];
+    return content;
+  }
+
+  /**
+   * Get the next item in the console history.
+   *
+   * @param activeCell - The currently selected Cell in the notebook.
+   * @param update - the promise returned from back or forward
+   */
+  updateEditor(activeCell: Cell, update: string): void {
+    if (activeCell) {
+      const model = activeCell.editor?.model;
+      const source = model?.sharedModel.getSource();
+      if (this.isDisposed || !update) {
+        return;
+      }
+      if (source === update) {
+        return;
+      }
+      this._setByHistory = true;
+      model?.sharedModel.setSource(update);
+      let columnPos = 0;
+      columnPos = update.indexOf('\n');
+      if (columnPos < 0) {
+        columnPos = update.length;
+      }
+      activeCell.editor?.setCursorPosition({ line: 0, column: columnPos });
+    }
+  }
+
+  /**
+   * Reset the history navigation state, i.e., start a new history session.
+   */
+  reset(): void {
+    this._hasSession = false;
+    this._placeholder = '';
+  }
+
+  /**
+   * Populate the history collection on history reply from a kernel.
+   *
+   * @param value The kernel message history reply.
+   *
+   * #### Notes
+   * History entries have the shape:
+   * [session: number, line: number, input: string]
+   * Contiguous duplicates are stripped out of the API response.
+   */
+  protected onHistory(
+    value: KernelMessage.IHistoryReplyMsg,
+    cell?: Cell
+  ): void {
+    this._history.length = 0;
+    let last = ['', '', ''];
+    let current = ['', '', ''];
+    let kernelSession = '';
+    if (value.content.status === 'ok') {
+      for (let i = 0; i < value.content.history.length; i++) {
+        current = value.content.history[i] as string[];
+        if (current !== last) {
+          kernelSession = (value.content.history[i] as string[])[0];
+          this._history.push((last = current));
+        }
+      }
+      // set the kernel session for filtering
+      if (!this.kernelSession) {
+        if (current[2] == cell?.model.sharedModel.getSource()) {
+          this._kernelSession = kernelSession;
+        }
+      }
+    }
+  }
+
+  /**
+   * Handle a text change signal from the editor.
+   */
+  protected onTextChange(): void {
+    if (this._setByHistory) {
+      this._setByHistory = false;
+      return;
+    }
+    this.reset();
+  }
+
+  /**
+   * Handle the current kernel changing.
+   */
+  private async _handleKernel(): Promise<void> {
+    this._kernel = this.sessionContext.session?.kernel;
+    if (!this._kernel) {
+      this._history.length = 0;
+      return;
+    }
+    await this._retrieveHistory().catch();
+    return;
+  }
+
+  /**
+   * handle a cell execution
+   */
+  private async _onExecuted(
+    _: unknown,
+    args: { notebook: StaticNotebook; cell: Cell }
+  ): Promise<void> {
+    const cell = args['cell'];
+    await this._retrieveHistory(cell).catch();
```

Comment:
This is still fetching 500 entries of history on each cell execution. I just tried this locally and for my current history it is extra 180kB on the websocket wire for each cell; execute a notebook with 100 cells and you got 18MB noise on websocket. This is compared to execution and reply messages being barely 300-600 bytes. 

I would suggest we only fetch history once when it is first requested by the user.

Reply:
I have it fetching history on each reset so that newly run cells are included in the history without manually appending them. Also, fetching it each time allows a user to access the history from different kernels as well, though they would have to scroll through the history of the current kernel first. What if I implemented something where it only requested a few at a time, requesting more as the user scrolls farther back through the history.

I'm also now wondering if accessing different kernel histories could be considered a security risk with multiple users on jupyter_server? Would users be able to access the kernel history for other users as it is now?

Reply:
It is fine to fetch say 10-20 items at a time and then maybe more if there is no overlap with what was fetched before. All I am arguing is that it should never happen on cell execution, but only when user does invoke `notebook:access-previous-history-entry`/`notebook:access-next-history-entry`. We could use cell execution as a heuristic for how many items to fetch; in pseudo-code:

```ts
class {
  onCellExecuted() {
     this._minToFetch += 1;
  }
  
  previousEntry() {
    this._ensureHistorySynced();
    this._setPrevious()
  }
  
  private _ensureHistorySynced() {
    let historyReconcliliated = false;
    let initialToFetch = this._minToFetch + 5;
    let historyChunk = this._fetch(initialToFetch);
    const historyChunks = [];
    while (!historyReconcliliated) {
      if (historyChunk overlaps this._historyCache) {
        historyChunks.push(historyChunk[overlapStart:]);
        historyReconcliliated = true;
      } else {
        // this assumes no execution in the meantime
        historyChunks.push(historyChunk);
        historyChunk = this._fetch(10);
      }
    }
    this._historyCache.push(...reversed(historyChunks));
    this._minToFetch = 0;
  }
  private _minToFetch = 0;
}
```

Though this will be tricky to get right. Do we really want to have history from concurrently executing kernels though?

Maybe using `session`, `start` and `stop` arguments of, `history_request` could help? In particular if we restrict ourselves to single `session`, the problem of cross-session history leakage would be eliminated (but then the history would be more limited).

Reply:
Yeah I was on the fence on whether or not to include multiple kernel histories or just the history of the notebook kernel, I ultimately decided to include them all since that's how the console history works. I did mostly implement the single notebook kernel history though, and that's actually why cell history is collected on execution. I couldn't figure out another way to get the kernel id number without collecting from the first history item immediately after a cell in the notebook was run, though if we go that way it should probably only call history after the first execution. So it seems to me the options are:

1) include whole history, and only collect on access commands or
2) include only history from current kernel, getting kernel number from first execution of the session

and then with either of those options it can either:

a) grab history in intervals or
b) grab n number (what we currently have)

What are your thoughts?

Reply:
@rsaim, do you have an opinion on this?